File: webforj-components/webforj-html-elements/src/main/java/com/webforj/component/html/elements/OrderedList.java
package com.webforj.component.html.elements;

import com.webforj.component.Component;
import com.webforj.component.element.annotation.NodeName;
import com.webforj.component.html.HtmlComponentContainer;

/**
 * Component representing a {@code ol} element.
 *
 * @see <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ol">HTML ol Tag</a>
 *
 * @author Hyyan Abo Fakher
 * @since 23.06
 */
// We're purposefully ignoring the deep inheritance warning here because we've designed our class
// hierarchy to meet the unique requirements of our UI framework. This design closely aligns with
// our framework's specific goals and emphasizes the need for caution when considering any changes.
//
// Any changes to the inheritance structure should be thoughtfully evaluated in the context of our
// framework's needs. The current structure is essential for meeting those needs.
@SuppressWarnings("squid:S110")
@NodeName("ol")
public class OrderedList extends HtmlComponentContainer<OrderedList> {

  /**
   * Creates a new empty OrderedList.
   */
  public OrderedList() {
    super();
  }

  /**
   * Creates a new OrderedList with the given text.
   *
   * @param text the text
   */
  public OrderedList(String text) {
    super();
    setText(text);
  }

  /**
   * Creates a new OrderedList with the given child components.
   *
   * @param components the child components
   */
  public OrderedList(Component... components) {
    super(components);
  }
}

File: webforj-kotlin/src/main/kotlin/com/webforj/kotlin/dsl/component/html/elements/OrderedList.kt
package com.webforj.kotlin.dsl.component.html.elements

import com.webforj.component.html.elements.ListEntry
import com.webforj.component.html.elements.OrderedList
import com.webforj.concern.HasComponents
import com.webforj.kotlin.dsl.WebforjDsl
import com.webforj.kotlin.dsl.init

/**
 * Creates an `OrderedList` with an optional [text].
 * ```
 * ... {
 *   orderedList() // Empty ol element
 *   orderedList("text") // ol element with text
 * ```
 *
 * @param text The text to add to the `OrderedList`.
 * @param block The initialization steps for the `OrderedList`.
 * @return The configured `OrderedList`.
 * @see OrderedList
 * @see <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ol">HTML ol Tag</a>
 */
fun @WebforjDsl HasComponents.orderedList(text: String? = null, block: @WebforjDsl OrderedList.() -> Unit = {}): OrderedList {
    val ul = text?.let { OrderedList(text) } ?: OrderedList()
    return init(ul, block)
}

/**
 * Creates an `ListEntry` for an [OrderedList] with an optional [text].
 * ```
 * orderedList {
 *   listEntry() // Empty li element
 *   listEntry("text") // li element with text
 * }
 * ```
 *
 * @param text The text to add to the `ListEntry`.
 * @param block The initialization steps for the `ListEntry`.
 * @return The configured `ListEntry` for the [OrderedList].
 * @see ListEntry
 * @see <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/li">HTML li Tag</a>
 */
fun @WebforjDsl OrderedList.listEntry(text: String? = null, block: @WebforjDsl ListEntry.() -> Unit = {}): ListEntry {
    val li = text?.let { ListEntry(text) } ?: ListEntry()
    return init(li, block)
}
